floors/floors.go

```diff
@@ -0,0 +1,203 @@
+}
+
+// resolveFloors does selection of floors fields from request JSON and dynamic fetched floors JSON if dynamic fetch is enabled
+func resolveFloors(account config.Account, bidRequestWrapper *openrtb_ext.RequestWrapper, conversions currency.Conversions) (*openrtb_ext.PriceFloorRules, []error) {
+	var errList []error
+	var floorsJson *openrtb_ext.PriceFloorRules
+
+	reqFloor := extractFloorsFromRequest(bidRequestWrapper)
+	if reqFloor != nil {
+		floorsJson, errList = createFloorsFrom(reqFloor, openrtb_ext.FetchNone, openrtb_ext.RequestLocation)
+	} else {
+		floorsJson, errList = createFloorsFrom(nil, openrtb_ext.FetchNone, openrtb_ext.NoDataLocation)
+	}
+	return floorsJson, errList
+}
+
+// createFloorsFrom does preparation of floors data which shall be used for further processing
+func createFloorsFrom(floors *openrtb_ext.PriceFloorRules, fetchStatus, floorLocation string) (*openrtb_ext.PriceFloorRules, []error) {
+	var floorModelErrList []error
+	finalFloors := new(openrtb_ext.PriceFloorRules)
+
+	if floors != nil {
+		floorValidationErr := validateFloorParams(floors)
+		if floorValidationErr != nil {
+			finalFloors.FetchStatus = fetchStatus
+			finalFloors.PriceFloorLocation = floorLocation
+			return finalFloors, append(floorModelErrList, floorValidationErr)
+		}
+
+		finalFloors.Enforcement = floors.Enforcement
+		if floors.Data != nil {
+			validModelGroups, floorModelErrList := selectValidFloorModelGroups(floors.Data.ModelGroups)
+			if len(validModelGroups) == 0 {
+				finalFloors.FetchStatus = fetchStatus
+				finalFloors.PriceFloorLocation = floorLocation
+				return finalFloors, floorModelErrList
+			} else {
+				*finalFloors = *floors
+				finalFloors.Data = new(openrtb_ext.PriceFloorData)
+				*finalFloors.Data = *floors.Data
+				if len(validModelGroups) > 1 {
+					validModelGroups = selectFloorModelGroup(validModelGroups, rand.Intn)
+				}
+				finalFloors.Data.ModelGroups = []openrtb_ext.PriceFloorModelGroup{validModelGroups[0].Copy()}
+			}
+		}
+	}
+	finalFloors.FetchStatus = fetchStatus
+	finalFloors.PriceFloorLocation = floorLocation
+	return finalFloors, floorModelErrList
+}
+
+// extractFloorsFromRequest gets floors data from req.ext.prebid.floors
+func extractFloorsFromRequest(bidRequestWrapper *openrtb_ext.RequestWrapper) *openrtb_ext.PriceFloorRules {
+	requestExt, err := bidRequestWrapper.GetRequestExt()
+	if err == nil {
+		prebidExt := requestExt.GetPrebid()
+		if prebidExt != nil && prebidExt.Floors != nil {
+			return prebidExt.Floors
+		}
+	}
+	return nil
+}
+
+// updateFloorsInRequest updates req.ext.prebid.floors with floors data
+func updateFloorsInRequest(bidRequestWrapper *openrtb_ext.RequestWrapper, priceFloors *openrtb_ext.PriceFloorRules) {
+	requestExt, err := bidRequestWrapper.GetRequestExt()
+	if err == nil {
+		prebidExt := requestExt.GetPrebid()
+		if prebidExt != nil {
+			prebidExt.Floors = priceFloors
+			requestExt.SetPrebid(prebidExt)
+			bidRequestWrapper.RebuildRequestExt()
```

Comment:
Why are we doing a rebuild at this point? Is there some code further along that is accessing the raw request rather than the wrapper and thus needs to see the modified values in the raw request? If so, then I would prefer fixing that code to use the wrapper rather than refactoring the wrapper to work around that deficiency in the code. Or at worst, moving the rebuild to the start of the code that depends on the raw request being up to date so it is only rebuilt if and when necessary.

Reply:
As mentioned in  [Floors Signalling Section](https://docs.google.com/document/d/1YAt00RfQaU2Ugciqw1LnirLgjyUz9ibfyjfxWUHR_P0/edit#bookmark=id.v4euxvet2sjh), only selected model groups are kept into req.ext.prebid.floors

Reply:
I don't think I was clear here. I am not asking why the request `ext` fields need to be updated. I am asking why the changes need to be written back to the raw JSON in the original openrtb struct. All changes should be able to live in the wrapper until it is time to marshal the openrtb request. Or possibly because we need to access the updated `ext` as a JSON blob rather than in the wrapper format. I don't see you doing either of these two things here, so either that points to the existence of some other code in the project which isn't properly using the wrapper, or we don't need to issue the Rebuild here. If there is some other code someplace that isn't properly using the wrapper, I want to get that fixed.

The purpose of the wrapper is to defer all marshaling of these `ext` fields until the very end. Each unmarshal/remarshal sequence is expensive, so it is preferable to only do it once per request. By forcing a rebuild here, any downstream code that also works with these `ext`s will require a second unmarshal/remarshal round. So I would like to remove this, or understand why the wrapper has failed in its design goal.

Reply:
The reason behind updating request extension is to update resolved request JSON with updated floors JSON in debug = true case.

Reply:
added full rebuild